File: python/icatt/elements.py
import numpy as np
from numba import njit, jit
import time

def kepler_elements(times):
    r = np.array([8.59072560e+02, -4.13720368e+03, 5.29556871e+03])
    v = np.array([7.37289205e+00, 2.08223573e+00, 4.39999794e-01])
    mu = 3.986004418e5
    el = elements(r, v, mu)

    best = np.inf
    worst = -np.inf
    total = 0.0
    for _ in range(times):
        t0 = time.clock()
        el = elements(r, v, mu)
        t1 = time.clock()
        current = t1 - t0
        if current < best:
            best = current
        if current > worst:
            worst = current
        total += current
    print("[",total/times,",",best,",",worst,"]")

@njit
def elements(r, v, mu):
    k = np.zeros(3)
    k[2] = 1.0
    el = np.empty(6)
    r_mag = np.dot(r, r)**(1/2)
    v_mag = np.dot(v, v)**(1/2)
    h = cross(r, v)
    h_mag = np.dot(h, h)**(1/2)
    n = cross(k, h)
    n_mag = np.dot(n, n)**(1/2)
    xi = v_mag ** 2 / 2 - mu / r_mag
    e = ((v_mag ** 2 - mu / r_mag) * r - v * np.dot(r, v)) / mu
    el[1] = np.dot(e, e)**(1/2)
    if not el[1] == 1:
        el[0] = - mu / (2 * xi)
    else:
        el[0] = h_mag ** 2 / mu
    el[2] = np.arccos(h[2] / h_mag)
    el[3] = np.arctan2(n[1]/h_mag, n[0]/h_mag)
    el[4] = np.arccos(np.dot(n, e) / (el[1] * n_mag))
    el[5] = np.arccos(np.dot(e, r) / (el[1] * r_mag))
    # Quadrant checks
    if n[1] < 0:
        el[3] = 2*np.pi - el[3]
    if e[2] < 0:
        el[4] = 2*np.pi - el[4]
    if np.dot(r, v) < 0:
        el[5] = 2*np.pi - el[5]
    return el

@njit
def cross(a, b):
    c = np.zeros(3)
    c[0] = a[1] * b[2] - a[2] * b[1]
    c[1] = a[2] * b[0] - a[0] * b[2]
    c[2] = a[0] * b[1] - a[1] * b[0]
    return c

File: cpp/src/elements.cpp
#include <iostream>
#include <chrono>
#include <Eigen/Dense>
#include "elements.h"

using Eigen::Vector3d;

namespace elements
{
    VectorXd elements(Vector3d r, Vector3d v, double mu)
    {
        auto r_mag = r.norm();
        auto v_mag = v.norm();
        auto v_mag2 = v_mag * v_mag;
        auto h = r.cross(v);
        auto h_mag = h.norm();
        Vector3d k(0,0,1);
        auto n = k.cross(h);
        auto n_mag = n.norm();
        auto xi = v_mag2/2 - mu/r_mag;
        auto e = ((v_mag2-mu/r_mag)*r - v*r.dot(v))/mu;
        auto ecc = e.norm();
        double sma;
        if (ecc != 1) {
            sma = -mu/(2*xi);
        } else {
            sma = pow(h_mag,2)/mu;
        }
        auto inc = acos(h.z()/h_mag);
        auto node = acos(n.x()/n_mag);
        auto peri = acos(n.dot(e)/(ecc*n_mag));
        auto ano = acos(e.dot(r)/(ecc*r_mag));
        if (n.y() < 0) {
            node = M_PI*2 - node;
        }
        if (e.z() < 0) {
            peri = M_PI*2 - peri;
        }
        if (r.dot(v) < 0) {
            ano = M_PI*2 - ano;
        }
        VectorXd out(6); out << sma, ecc, inc, node, peri, ano;
        return out;
    }

    void benchmark(int times) {
        auto mu = 3.986004418e5;
        Vector3d r(8.59072560e+02, -4.13720368e+03, 5.29556871e+03);
        Vector3d v(7.37289205e+00, 2.08223573e+00, 4.39999794e-01);

        auto best = std::numeric_limits<double>::infinity();
        auto worst = -std::numeric_limits<double>::infinity();
        double all = 0;
        for (auto i=0; i < times; i++) {
            auto begin = std::chrono::high_resolution_clock::now();

            elements(r, v, mu);

            auto end = std::chrono::high_resolution_clock::now();
            auto current = std::chrono::duration_cast<std::chrono::nanoseconds>(end-begin).count()/1e9;
            all += current;
            if (current < best) {
                best = current;
            }
            if (current > worst) {
                worst = current;
            }
        }
        std::cout << "[" << all/times << "," << best << "," << worst << "]" << std::endl;
    }
}
